presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java

```diff
@@ -388,14 +423,52 @@ public void addInput(Page page)
             checkState(!aggregationBuilder.isFull(), "Aggregation buffer is full");
         }
 
+        if (!preGroupedChannels.isEmpty()) {
+            int groupEnd = findGroupEnd(0, page.getPositionCount(), page);
+            if (groupEnd < page.getPositionCount()) {
+                unfinishedWork = aggregationBuilder.processPage(page.getRegion(0, groupEnd));
+                pendingInput = page.getRegion(groupEnd, page.getPositionCount() - groupEnd);
+            }
+            else {
```

Comment:
if groups always span entire page, then you will never flush aggregates

Reply:
@sopel39 Good catch. Thank you, Karol.